support openssl ALPN extension

superisaac opened the pull request:
This pr is an implement of ALPN extension, It does't change currenty module too much.
To make alpn work the client needs to call `setAlpnProtocols()` and the server need to call `setAlpnSelectCallback()`, refer to tests/stdlib/thttpclient_ssl_alpn.nim

bung87 committed:
Shadow Dom apis (#14979)

* shadow dom api

* fix typos

* host to Element type

* fix code style

* move elementsFromPoint to dom_extensions.nim

genotrance committed:
Fix #2408 - add -d:globalSymbols (#14904)

narimiran committed:
fix several newline problems (#15028) [backend]

* prevent newlines where they shouldn't be
* 'contentLength' shouldn't be negative

Araq committed:
ARC: optimize the code better when --panics:off (#15031)

mildred committed:
smtp: Fix STARTTLS, request HELO once TLS is established (#15032)

Commit:
Add a testcase for #14480. Fixes #14480 (#15037)

Commit:
Add a testcase for #14480. Fixes #14480 (#15037)

HugoGranstrom committed:
fix #15033 (#15034)

Araq committed:
fixes #15026 [backport] (#15040)

no test case since only a special case was affected and the special case got removed

narimiran committed:
Change testing commands for some packages (#15041)

Araq committed:
enforce browsers.nim only handles URLs [backport] (#15045)

superisaac committed:
support openssl ALPN extension

superisaac committed:
support openssl ALPN extension

juancarlospaco commented on lib/pure/net.nim:
```diff
@@ -728,6 +728,15 @@ when defineSsl:
+    ## Returns the negotiaged ALPN protocols.
+    ## If the ALPN extension is not available or no ALPN candidate
+    ## provided by clients then returns empty string.
+    if not socket.isSsl:
```
The `result` is autoinit to `""` then

```nim
if socket.isSsl: result = socket.sslHandle.getAlpnProtocol()
```

superisaac replied:
thank you for your comment, the fix commit is pushed as a405eec1

juancarlospaco commented on lib/pure/asyncnet.nim:
```diff
@@ -772,6 +772,15 @@ when defineSsl:
+    ## Returns the negotiaged ALPN protocols.
+    ## If the ALPN extension is not available or no ALPN candidate
+    ## provided by clients then returns empty string.
+    if not socket.isSsl:
```
Same.

superisaac committed:
refs #15046, changed setAlpnCallback() -> setAlpnSelectCallback(); fixed comments; added comments; fixed unittests

superisaac committed:
refs #15046, changed setAlpnCallback() -> setAlpnSelectCallback(); fixed comments; added comments; fixed unittests

superisaac committed:
Merge branch 'devel' into ssl-alpn

Araq committed:
fixes #15056 [backport]

narimiran committed:
jsre: try to fix nightlies (#15057)

jcosborn committed:
fix assignment to converted concept type (#15051)

* fix assignment to converted concept type

* check for resolved concepts

* add extra test

genotrance committed:
Bump nimble commit (#15053)

bung87 committed:
fix #14534 (#15060) [backport]

bung87 committed:
fix #14684 (#15059)

narimiran committed:
deprecate tables.add (#15047)

Araq committed:
writing to a location counts as "side effect"; implements https://github.com/nim-lang/RFCs/issues/234 (#15030)

Araq committed:
strict func: much better error messages (#15068)

* strict func: much better error messages
* documented the 'strict funcs' mode

Araq committed:
fixes #15052

Commit:
Add a test-case for #12990 (#15072)

* closes #12990
* Add a test-case for #12990

Araq committed:
fixes #15038 [backport:1.2]

alaviss committed:
koch: bundle nim-lang/fusion with Nim (#15061)

Initial work on bundling nim-lang/fusion as part of the Nim distribution.

narimiran committed:
fix nightlies: smaller log files (#15074)

Hopefully this will fix
"The job exceeded the maximum log length, and has been terminated."

genotrance committed:
Bump nimble (#15077)

Araq committed:
compiler: minor code cleanups

Jjp137 committed:
parsecsv: fix '\0' being displayed as '0' in docs (#15086) [backport]

alaviss committed:
Small optimization for the CI pipeline. (#15088)

* azure-pipelines: cache csources to speed up build

Building csources takes about 2-4 mins, and since it hasn't changed for
a long time, employ some caching to cut the time spent.

* azure-pipelines: remove ppa avoidance configs

Testing shows that there aren't any difference between having this and
not having it, so removing this to simplify the code.

Araq committed:
fixes a closure iterator memory leaks, progress on #15076

Araq committed:
code cleanup

Araq committed:
more renamings

Commit:
Add test-cases for #12576 and #12523 (#15085)

* Add a test-case for #12576

* Add a test-case for #12523

bung87 committed:
fixes #14189 (#15080) [backport]

superisaac committed:
resolve conflict with not owned file

superisaac committed:
Merge branch 'devel' into ssl-alpn

alaviss commented:
/cc @FedericoCeratto
